### PR TITLE
[B+C] Added EntityPotionEffectChangeEvent. Adds BUKKIT-2817

### PR DESCRIPTION
**Pull request by turt2live**
**The Issue:**

There is no current way to handle when an entity (especially players) gain or lose a potion effect.

**Justification**

With this addition, plugins will be able to control further what potion effects players can have, or cannot have. This could be used within plot plugins to restrict beacons having an effect outside of the placed plot.

**PR Breakdown:**

This PR adds a new event for plugins to use. Within CraftBukkit there is implementation to handle gain/loss of effects.

**Testing Results / Materials:**

Plugin code used to test: https://gist.github.com/turt2live/5369425
Compiled Binary: http://home.turt2live.com/PotionTest.jar

Steps to apply cancel testing, type "/test". To revert, type "/test" again.

To apply a plugin-made effect, type "/test a"

**Associated PRs**

B-717 - https://github.com/Bukkit/Bukkit/pull/717 - API portion for this event
CB-967 - https://github.com/Bukkit/CraftBukkit/pull/1242 - Implementation for this event

**JIRA Tickets**

Main ticket: https://bukkit.atlassian.net/browse/BUKKIT-2817
Subsequent (duplicate) tickets
https://bukkit.atlassian.net/browse/BUKKIT-2361 
https://bukkit.atlassian.net/browse/BUKKIT-1535 
https://bukkit.atlassian.net/browse/BUKKIT-1778 
https://bukkit.atlassian.net/browse/BUKKIT-763

### CODE REVIEW

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+     * @param ambient true if ambient
+     * @param location the location of the event, if applicable
+     */
+    public EntityPotionEffectChangeEvent(LivingEntity who, PotionEffect effect, boolean lostEffect, PotionChangeCause cause, boolean ambient, Location location){
+        super(who);
+        this.lostEffect = lostEffect;
+        this.effect = effect;
+        this.cause = cause;
+        this.ambient = ambient;
+        this.location=location.clone();
```
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Style - spaces around `=`

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
```
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Style - excess spaces

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    public boolean isCancelled(){
+        return cancel;
+    }
+
+    /**
+     * Set the event as cancelled.<br>
+     * The event will only cancel if the entity is <b>not</b> losing the effect.
+     *
```
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Style - Excess space

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    }
+    
+    /**
+     * Gets the ambient flag for beacons. This will be false otherwise (unless set by a plugin)
+     * @return true if ambient
+     */
+    public boolean isAmbient(){
+        return ambient;
+    }
+
```
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Style - Excess spaces

**Comment by riking** [comment]
Please update the whole file to current style guide

**Comment by turt2live** [comment]
Will fix as well (when I get home..)

**Comment by turt2live** [comment]
@riking  Fixed this half

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+    @Override
+    public HandlerList getHandlers(){
+        return handlers;
+    }
+
+    public static HandlerList getHandlerList(){
+        return handlers;
+    }
+}
\ No newline at end of file
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Needs a newline

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    public Location getLocation(){
+        return location;
+    }
+
+    /**
+     * Gets the ambient flag for beacons. This will be false otherwise 
+     * (unless set by a plugin)
+     * @return true if ambient
+     */
+    public boolean isAmbient(){
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]:
``` java

isAmbient() {

```

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: ... and all subsequent. On it.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+     * effect.
+     * @return the potion effect
+     */
+    public PotionEffect getEffect() {
+        return effect;
+    }
+
+    /**
+     * Used to determine if the entity is going to lose the effect
+     * @return true if the entity will lose the effect, false otherwise
```
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: This seems like a double negative in nature to me.  Consider `isGainingEffect` instead.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+ * Fired when a entity's potion effects change. <br>
+ * Causes for firing:<br>
+ * <ul>
+ * <li>Beacon potion effect gained/lost</li>
+ * <li>Splash/edible potion effect gained/lost</li>
+ * <li>Plugin altered potion effect gained/lost</li>
+ * <li>Mob (Cave Spider, Zombie, Wither, etc) attacks entity</li>
+ * </ul>
+ */
+public class EntityPotionEffectChangeEvent extends EntityEvent implements Cancellable {
```
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:24]: Have you considered `EntityEffectChangeEvent`?  I personally think Bukkit's naming of `PotionEffect` is misleading, especially in light of how they can be applied from other events that don't include potions.  Even nms refers to them as `MobEffect`s.

I know it'd be a bit of a inconsistency in naming from a Bukkit perspective, but I'd almost like to see `PotionEffect` be the inconsistent name over time and everything else migrate to just `Effect`.  This is a personal preference, I can't guarantee the core team would agree at this point.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+    public static enum PotionChangeCause {
```
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Would this be more consistent as `PotionEfffectChangeCause`?  Or even just `Cause` actually since it's within the class that defines what it relates to and even external references could refer to it as `EntityPotionEffectChangeEvent.Cause` if clarity outside of local variable naming was desired.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        /**
+         * Wither-like entity caused event
```
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: "Wither-like"?  You mean wither skeleton or wither?  Why not specify each specifically with their own enum entry?

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+
+        /**
+         * Plugin altered effect
+         */
+        PLUGIN,
+
+        /**
+         * Effect loss
+         */
+        LOSS;
```
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: If you have `isLosingEffect` why would you need this cause or vice versa?

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 8 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:24]: I would stick with Potion Effect to differentiate between a potion (or a mob effect) from a visual effect (TNT explosion)

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 9 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Agreed, will change when I get home

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 10 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: It's the cause of the wither effect being applied, I'd argue that the JavaDoc is badly written over the cause enum.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 11 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Mostly for simplicity in the construction of the event, but I guess a second constructor and a second method in CraftEventFactory could work.

EDIT: Upon further review, this would also mean that the cause will be null, and that to me doesn't seem ideal

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 7 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Sure

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 8 above)
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:24]: Oh you are right, I forgot about `org.bukkit.Effect` which covers sound and visual effects. :(

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 10 above)
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Well the effect would be determined from `getEffect`.  Isn't the Cause enum supposed to provider further clarity as to the source of the effect being applied and not just duplicate available event information?

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 11 above)
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I'm not sure I follow.  While I haven't analyzed the CraftBukkit PR to determine why it's like this, I'm simply assessing the Javadoc information and the API structure at this point.  A cause of LOSS clarifies no source or reason as to why the change is occurring and seems to go against the purpose of the Cause enum.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I'm not a fan of having a null cause though, if the Bukkit team however wants that I will do it, but imo event.getCause() == null is somewhat not helpful.

Maybe replacing LOSS with UNKNOWN to represent a not-null cause? 

The reason why I don't like null causes is so I can do something like this (just my opinion of course):

``` java
if(event.getCause() == SOME CAUSE){
    // Do stuff
}
```

The above code would throw (iirc) a NPE.

Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: No, it'd only throw NPE if event was null. It'd be perfectly valid to do

``` java
if (event.getCause() == null) {
    // Do stuff
}
```

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: My question more is to the point of why would you even need a null or no-op value here?  Why would an effect change without a reason that can be identified?

Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: @EdGruberman The reason is already identfied - loss of the effect.
The reason for a null value is not having a LOSS enum available.
Or do you mean something more like EXPIRED (time), MILK (bucked of milk), and DEATH as "loss" causes?

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: EXPIRED and MILK (maybe TREATED instead?), yes!  Those are causes.  But leaving it a generic LOSS when you take the time to isolate such specific causes as CAVE_SPIDER just did not seem consistent.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I'll change it to EXPIRED then?

And a null cause would be due to the contract of the constructor - A cause

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: As I said, EXPIRED sounds good for if you are inferring it will be used when this event is raised due to an effect timing out.  But what about using milk?  I almost think a TREATED will be needed also.

I still don't know what you mean by null being used.  Are you just recognizing that someone could call the event with a null cause?  That's fine.  But I don't see why a CraftBukkit implementation would do that.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 10 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I could limit it down to MOB, BEACON, POTION, PLUGIN, EXPIRE/Whatever. Then there is plenty of clarity.

Or I could split it out to WITHER_BOSS and WITHER_SKELETON.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 11 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: REMOVED would fit both descriptions.

In order for me to follow the minimal diffs I used a spot in CB where the overall "loss" of the effect (due to whatever cause). If you want I could track it down to where all the possible remove causes are and use those instead. 

Because of the constructor of this event (I'm not too fond of a second constructor) the cause would be null where CB would fire a loss. Due to the above, the simple "removed" hook would have no explicit cause, so a null cause would need to be passed for a LOSS cause (assuming the LOSS enum is missing). 

TL;DR: 

REMOVED could work.
Minimal diffs = No "advanced" lost causes (milk, etc). Can fix if needed (add MILK, EXPIRED, etc).
The cause would be null if the CURRENT CB implementation was kept and the LOSS enum removed.

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: REMOVED is just a synonym for LOSS.  I see no advantage to using it, it demonstrates no cause.  Using null instead of an explicit cause is not ideal either.  It'd be better to use LOSS or REMOVED as a null object pattern type of approach if necessary.

EXPIRED indicates the time ran out.  TREATED could indicate some form of counter acting agent was applied.  I'm not sure why this is hard for me explain.  A cause is something that explains why the event is occurring.  It is not a generalized state attribute.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I'm more or less asking this:

Do you want EXPIRED/TREATED or do you want less diffs?

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 10 above)
Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: MOB is an interesting concept... that almost seems to fit the enum better actually.  I suppose we'd need to dream up use cases to understand what might be best to put here all said and done.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Well, as you said, you can pretty much determine what hit the entity by reading the effect (a wither effect will not be applied by a zombie), so I guess MOB will make it simpler in the idea that `switch(cause){ case ZOMBIE: case WITHER: // etc }` is not needed.

**Comment by riking** [comment]
The only CB diff you need for TREATED is this:

``` diff
@ src/main/java/net/minecraft/server/EntityLiving.java
@@ -1768,1768 +1768,1769 @@
    public void bB() {
        Iterator iterator = this.effects.keySet().iterator();

        while (iterator.hasNext()) {
            Integer integer = (Integer) iterator.next();
            MobEffect mobeffect = (MobEffect) this.effects.get(integer);

            if (!this.world.isStatic) {
+              CraftEventFactory.callEntityPotionEffectChangeEvent(this, true, mobeffect, org.bukkit.event.entity.EntityPotionEffectChangeEvent.PotionChangeCause.TREATED, world, locX, locY, locZ); // CraftBukkit
                iterator.remove();
                this.c(mobeffect);
            }
        }
    }
```

**Comment by turt2live** [comment]
I'll have to play around with it. 

And that will have to happen later on.

Thanks for the input

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 11 above)
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Yes, split it to `EXPIRED` and `TREATED`.

Maybe also `PLUGIN` -> `PLUGIN_ADDED` / `PLUGIN_REMOVED`?

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Sure, thoughts on PLUGIN_GIVE and PLUGIN_TAKE?

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: On another note: BEACON_GIVE, BEACON_TAKE?

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: And on yet another note, EntityZombie calls EntityLiving.o(int id); to remove an effect (WEAKNESS specifically), would that be EXPIRED or TREATED?

Nothing else appears to call this method.

Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Beacons don't remove the effect, they just let it expire.

Kind of a bad way to do it, but hey, they don't have PlayerMoveEvent to work with :3

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Ah, right.

**Comment by turt2live** [comment]
Updated PR

**Comment by riking** [comment]
> And on yet another note, EntityZombie calls EntityLiving.o(int id); to remove an effect (WEAKNESS specifically), would that be EXPIRED or TREATED?

Reposting this because it'll get lost in the outdated diff.

...Do they call that when they turn into villagers?

**Comment by turt2live** [comment]
In the latest commit it was changed to TREATED as from the looks of it, when a zombie does something it "treats" the weakness. Open to suggestions however

**Comment by riking** [comment]
Under what exact circumstances is that code executed?

**Comment by turt2live** [comment]
From a quick look (something about Golden Apples in hand) and the Minecraft Wiki:

Curing Villager Zombies.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+         */
+        PLUGIN_REMOVED,
+
+        /**
+         * The effect expired (timer reached 0:00)
+         */
+        EXPIRED,
+
+        /**
+         * The effect was treated (The entity used a milk bucket)
```
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Hmm... also put in the example of.. I dunno

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: The more I think on this, the more even TREATED seems like just another synonym for LOSS.

What about ITEM?  Then the Javadoc could perhaps be `Result of an item being used`.  This would also encompass POTION.  I'm not sure those really need to be split out anyways?

For that matter, I just looked more and now see PLUGIN_ADDED and PLUGIN_REMOVED.  Same concern there in that _ADDED and _REMOVED tell you already what ``isGainingEffect` should.  That should just be PLUGIN.

That'd leave MOB, BEACON, ITEM, PLUGIN, EXPIRED.  Seems basic and straight forward.

And did I miss a change somewhere along the way or has `isGainingEffect` always been based on the cause value?  I thought it was passed in for the constructor before?  That makes a lot more sense to me.  If you are saying you want the cause enum to indicate gain or loss of the effect, then we are working backwards here in regards to simplifying the enum.

Ideally I think you would want to pass in the gain/loss boolean at construction and have the cause enum clarify the reason.  Otherwise it's just a convenience method to interpret the overly complex enum.

Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: POTION is used for splash potions.

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Ah good point, perhaps change that to SPLASH or AREA?  AREA might be more future proof.  And that Javadoc indicates consumption applies, but my suggestions would put that into my ITEM suggestion.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Someone decide. Let me know when you've come to a decision. 

I'm getting tired of running around through the code trying to please the both of you.

Talk it over and decide.

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: PRs are for talking it over and deciding. I apologize for not having a more consistent message from the team, but we figure most times some type of feedback immediately is better than waiting until some much later time that this discussion all happens internally.

Besides, If the PR author can't figure out a solution that works for everyone based on feedback, what is the benefit to using the PR code?  The core team would just write it themselves.  Most of the problem is usually figuring out the best way to do something.  Coding it is the easy part.  If you don't want to help determine the best way to do this, you can always close this PR and let everyone know someone else should pick up the torch.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: When I submitted it I thought it was fine. After I updated it for the change in standards, I thought it was fine.

When I was told it was going to be pulled, I thought it was fine.

I'm really just trying to fix a bug that has been in my plugin for nearly 5 months now. The last minute changes on what to name an enum seem somewhat not needed.

Here are my suggestions for the problems:
1) Remove the isGaining method to avoid extra information (considering it just reads what enum value was passed anyways)
2) In regards to POTION - It is used for both consumed and splash potions. If I'm not mistaken there is a way already in the API to determine whether or not it's a splash (if not, POTION_SPLASH and POTION_CONSUMED are my suggestions)

I'm open to opinions, but I'm really looking to fix that bug in my plugin.

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: The issue is that we can't just think about what you or I personally want.  We have to think about the impact globally.  This is a good sized API feature addition.  If we just slam some random enum names in to place and they end up not making sense later, we can get cornered into a common law type of implementation we can't change.  It's important to spend the time to ensure it makes sense up front.  There are plenty of events already implemented that everyone just groans at because no upfront thought was put in to them.  We are trying to prevent adding any more examples to that situation.

Your personal motivation may be a bugfix for your plugin, but the impact of this PR is greater than just you.  All this feedback is not really "last minute".  PRs have been poorly managed to date and Bukkit has empowered some dedicated people to working through them.  All this discussion is the result.  It's a definite increase of feedback suddenly, but this could still take some time to hash out.

Deciding on what the enum should represent should really be driven by use cases.  So let's explore another option, why have a cause at all?  Leave it off for now and it can be added later if needed.  Keep isGaining instead.  If a plugin needs to determine cause they can monitor other events and deduce accordingly?

I was hoping you'd clarify why you thought a cause enum was needed.  I'm having trouble seeing a purpose and without one the enum could be defined any number of ways for any number of  guessed at reasons.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Well for instance my plugin in particular only cares about effects given by beacons.

Even if I didn't need just beacons, I'd say leave the cause enum in due to the fact that if it would get added later, why not add it now? 

Any particular comments on my suggestions?

Comment (EdGruberman) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Are you reading what I'm saying?  If you add it now, we have to discuss what other items to put in and why.  This will be a lot more discussion.  I think this is just the tip of the iceberg.  Without a clear reason for the other items in the enum, why not just put BEACON and OTHER?  Then as use cases come up we can simply append the enum appropriately?

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: I have read it, and I'd rather get the long discussion over with so I can get a fully functioning event into the API. If I wanted a partial event I would have removed the cause enum and said "Removed enum". 

If we were to have BEACON and OTHER then how would other plugins determine if a zombie hit a villager, if a player ate a potion, or if a wither applied an effect? 

Some use cases I can think of: 
- Replacing wither effect with another effect (strength for example) based on number of players attacking the thing ("party bonus"). 
- Applying a strength effect to an affected villager to make them slightly harder. 
- "Random Potions" (a renamed potion that applies a random effect)

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+         * The effect expired (timer reached 0:00)
+         */
+        EXPIRED,
+
+        /**
+         * The effect was treated. Current examples: <br>
+         * <li>Player used milk bucket</li>
+         * <li>Villager cure process was started</li>
+         */
+        TREATED;
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:74]: @riking  better?

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+        /**
+         * Caused by a potion, splashed or consumed
+         */
+        POTION,
+
+        /**
+         * Caused by an enchanted golden apple
+         */
+        ENCHANTED_GOLDEN_APPLE,
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:49]: This seems a bit long, open to suggestions

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+
+        /**
+         * Caused by a beacon block
+         */
+        BEACON,
+
+        /**
+         * Caused by a potion, splashed or consumed
+         */
+        POTION,
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:39]: It may be worth noting that MC applies a very quick potion effect to the player when a potion is consumed. To determine if the effect applied is a splash potion I would do this:

``` java
if(cause == POTION){
    if(duration > 0){
        // Is splash
    }
}
```

**Comment by turt2live** [comment]
Test materials updated (Both CB and Bukkit)

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+
+import org.bukkit.Location;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+import org.bukkit.event.entity.EntityEvent;
+import org.bukkit.potion.PotionEffect;
+
+/**
+ * Fired when a entity's potion effects change. <br>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: We use `<p>` for paragraph breaks in Javadocs.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+ * Causes for firing:<br>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: We use `<p>` for paragraph breaks in Javadocs.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+         */
+        PLUGIN_REMOVED,
+
+        /**
+         * The effect expired (timer reached 0:00)
+         */
+        EXPIRED,
+
+        /**
+         * The effect was treated. Current examples: <br>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: We use `<p>` for paragraph breaks in Javadocs.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+     * Gets the cause associated with this event
+     * @return the cause of the event
+     */
+    public Cause getCause() {
+        return cause;
+    }
+
+    /**
+     * Gets the potion effect applied to this event. Modifying this modifies
+     * the overall effect.<br>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: We use `<p>` for paragraph breaks in Javadocs.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 16 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: So as follows?

```
/**
 * <p>Fired when a entity's potion effects change.</p>
 * <p>Causes for firing:</p>
 * <ul>
 * <li>Beacon potion effect gained/lost</li>
 * <li>Splash/edible potion effect gained/lost</li>
 * <li>Plugin altered potion effect gained/lost</li>
 * <li>Mob (Cave Spider, Zombie, Wither, etc) attacks entity</li>
 * <li>The entity used a milk bucket to remove an effect</li>
 * </ul>
 */
```

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 20]
```diff
@@ -0,0 +1,171 @@
...
+            return false;
+        }
+    }
+
+    public boolean isCancelled() {
+        return cancel;
+    }
+
+    /**
+     * Set the event as cancelled.<br>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: We use `<p>` for paragraph breaks in Javadocs.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java (hunk 16 above)
Comment (riking) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: no.

```
/**
 * Fired when a entity's potion effects change.
 * <p>
 * Causes for firing:
 * <ul>
```

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: Okay then

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 21]
```diff
@@ -0,0 +1,177 @@
...
+
+        /**
+         * The effect expired (timer reached 0:00)
+         */
+        EXPIRED,
+
+        /**
+         * The effect was treated. Current examples:
+         * <p>
+         * <li>Player used milk bucket</li>
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java:70]: I see no `<ul>` and `</ul>` here.

File: src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+    public PotionEffect getEffect() {
+        return effect;
+    }
+
+    /**
+     * Used to determine if the entity is gaining the effect.
+     * @return true if the entity is gaining the effect, false otherwise
+     */
+    public boolean isGainingEffect() {
+        switch(cause){
```
Comment (gmcferrin) [inline src/main/java/org/bukkit/event/entity/EntityPotionEffectChangeEvent.java]: `switch (cause) {`

**Comment by turt2live** [comment]
See #926
